src/udf/src/external.rs

```diff
@@ -109,6 +119,22 @@ impl ArrowFlightUdfClient {
         }
     }
 
+    /// Call a function, retry up to 5 times / 3s if connection is broken.
+    pub async fn call_with_retry(&self, id: &str, input: RecordBatch) -> Result<RecordBatch> {
+        let mut backoff = Duration::from_millis(100);
+        for i in 0..5 {
+            match self.call(id, input.clone()).await {
+                Err(err) if err.is_connection_error() && i != 4 => {
+                    tracing::error!(%err, "UDF connection error. retry...");
+                }
+                ret => return ret,
+            }
+            tokio::time::sleep(backoff).await;
+            backoff *= 2;
+        }
+        unreachable!()
```

Comment:
If it's possible to fail for 5 times, IMO it should return `Err` here rather than `unreachable`?

Reply:
Yes. It will return at line 130 rather than the end of this function.

Reply:
Oh I'm blind...What about `&& i < 5`🤣